L26-Ability-to-restrict-max-threads-in-C++SyncServer.md

```diff
@@ -0,0 +1,93 @@
+* Author(s): Sree Kuchibhotla (sreecha)
+* Approver: vjpai
+* Status: In Review
+* Implemented in: C++
+* Last updated: June 6, 2018
+* Discussion at: 
+
+## Abstract
+C++ `ThreadManager` is a specialized thread pool used to implement the C++ Synchronous server. 
+
+Currently, `ThreadManager` can be configured to have minimum and maximum number of polling threads. While this helps to always have threads available to poll for work, it does nothing to throttle polling if the `ThreadManager` is overloaded with work.
+
+The proposal here is to add a notion of 'thread quota' to the `resource_quota` object. Currently, a resource quota object can be attached to the server (via `ServerBuilder::SetResourceQuota`). The idea here is to set a maximum number of threads on that resource quota object.
+
+Each of the `ThreadManager` objects in a given server check with the Server's resource quota object when creating new threads and will not create new threads if no quota is available.
+
+## Details
+More concretely, the following API changes are being proposed
+
+### 1. New *public* C++ API: `ServerBuilder::SetMaxThreads`
+ 
+```C++
+// File: include/grpcpp/resource_quota.h
+
+/// ResourceQuota represents a bound on memory and threads usage by the gRPC
+/// library. A ResourceQuota can be attached to a server (via \a ServerBuilder),
+/// or a client channel (via \a ChannelArguments).
+///
+/// gRPC will attempt to keep memory and threads used by all attached entities
+/// below the ResourceQuota bound.
+class ResourceQuota final : private GrpcLibraryCodegen {
+ public:
+  ...
+  /// Set the max number of threads that can be allocated from this
+  /// ResourceQuota object.
+  ///
+  /// If the new_max_threads value is smaller than the current value, no new
+  /// threads are allocated until the number of active threads fall below
+  /// new_max_threads. There is no time bound on when this may happen i.e none
+  /// of the current threads are forcefully destroyed and all threads run their
+  /// normal course.
+  ResourceQuota& SetMaxThreads(int new_max_threads);
```

Comment:
Is 0 a valid value for max threads? Is that how I say "I will never use this server synchronously. Do not create any threads. I'll be polling completion queues myself"?

Reply:
0 is a valid value but that will effectively make the server not respond to any incoming requests. So it is not a good idea to initialize max_threads to 0. However, if for whatever reason, the application suddenly decides to stop accepting incoming requests for sometime, it can set the max_threads to 0.

Reply:
btw, as I am thinking more about this, it might be easier to just implement option (1) at this point. This way, when applications resize the max_threads by calling `ResourceQuota::SetMaxThreads`, it is easier to make those changes visible to all thread managers..

Reply:
..and oh to answer your other question 
>>" Is that how I say "_I will never use this server synchronously. Do not create any threads. I'll be polling completion queues myself"_?" 

No, actually the server completion queues used by the sync server are *not* exposed to the application. So there is no question of the application trying to poll the queues itself.

If application wants to poll itself, it has to create its own completion queues an poll them ..

Reply:
Can you start the grpc discussion thread for this proposal? We should move extended comments there.

Reply:
Yes, I did: https://groups.google.com/forum/#!topic/grpc-io/YQQY0pGG9MI